### PR TITLE
ER-257 Generate vacancy number Job

### PR DESCRIPTION
**Pull request by leewadhams**
Currently it's a manual process to set up the reference data for the next vacancy number. There is a story in the next sprint to automate this. Instructions for the manual process are in Developer Setup on Confluence.

- Webjob is triggered by VacancyCreateEvent that is published to Storage Queue
- Uses a new collection that stores the last vacancy reference used and increments and assigns vacancy with new value.

Things to note:
- Single Storage Queue for all vacancy events that are published
- Single job will pick up all vacancy events and handle.
- Job now creates queue if it doesn't already exist.

### CODE REVIEW

File: src/Jobs/Recruit.Vacancies.Jobs/VacancyEvents/VacancyEventsJob.cs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        {
+            try
+            {
+                var eventItem = JsonConvert.DeserializeObject<EventItem>(message);
+                
+                await UnpackAndExecute(eventItem.EventType, eventItem.Data);
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(ex, $"Unable to run {JobName}");
```
Comment (mgwilliam) [inline src/Jobs/Recruit.Vacancies.Jobs/VacancyEvents/VacancyEventsJob.cs]: should this message be Unable to handle event?

File: src/Shared/Recruit.Vacancies.Client/Infrastructure/Events/StorageQueueEventQueue.cs [hunk 2]
```diff
@@ -31,14 +32,23 @@ public StorageQueueEventQueue(StorageQueueConnectionDetails details)
...
 
             await queue.AddMessageAsync(message);
         }
+
+        private static string GetQueueName(IEvent @event)
+        {
+            if (@event is IVacancyEvent)
+                return QueueNames.VacancyEventsQueueName;
+
+            return @event.GetType().Name.Replace("Event", "Queue").PascalToKebabCase();
```
Comment (mgwilliam) [inline src/Shared/Recruit.Vacancies.Client/Infrastructure/Events/StorageQueueEventQueue.cs]: seems unused - do we need this?

File: src/Shared/Recruit.Vacancies.Client/Application/CommandHandlers/AssignVacancyNumberCommandHandler.cs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+            if (vacancy.VacancyReference.HasValue)
+            {
+                _logger.LogWarning("Vacancy: {vacancyId} already has a vacancy number: {vacancyNumber}. Will not be changed.", vacancy.Id, vacancy.VacancyReference);
+            }
+
+            vacancy.VacancyReference = await _generator.GenerateAsync();
+
+            await _repository.UpdateAsync(vacancy);
+
+            _logger.LogInformation("Updated Vacancy: {vacancyId} with vacancy number: {vacancyNumber}", vacancy.Id, vacancy.VacancyReference);
```
Comment (mgwilliam) [inline src/Shared/Recruit.Vacancies.Client/Application/CommandHandlers/AssignVacancyNumberCommandHandler.cs]: missing $ maybe?

File: src/Shared/Recruit.Vacancies.Client/Application/CommandHandlers/AssignVacancyNumberCommandHandler.cs [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+            _generator = generator;
+        }
+
+        public async Task Handle(AssignVacancyNumberCommand message, CancellationToken cancellationToken)
+        {
+            var vacancy = await _repository.GetVacancyAsync(message.VacancyId);
+            
+            if (vacancy.VacancyReference.HasValue)
+            {
+                _logger.LogWarning("Vacancy: {vacancyId} already has a vacancy number: {vacancyNumber}. Will not be changed.", vacancy.Id, vacancy.VacancyReference);
```
Comment (mgwilliam) [inline src/Shared/Recruit.Vacancies.Client/Application/CommandHandlers/AssignVacancyNumberCommandHandler.cs]: missing $ maybe?

File: src/Shared/Recruit.Vacancies.Client/Application/CommandHandlers/AssignVacancyNumberCommandHandler.cs (hunk 3 above)
Comment (leewadhams) [reply src/Shared/Recruit.Vacancies.Client/Application/CommandHandlers/AssignVacancyNumberCommandHandler.cs]: Nope. Logger uses string format for structured logging. i.e. vacancyId and vacancyNumber will appear in Kibana as properties of the log message as well as in the message string itself.

**Comment by leewadhams** [comment]
Yeah IHandleVacancyEvent isn't really needed. Was originally thinking it was injected in to the job class. But then didn't see much of a reason to do it. I'll remove it.

**Comment by sachpatel** [comment]
Hi, I was able to run the job successfully. In the developer setup notes you mention creating a `vacancy-number` document with an initial value of 1. I think we wanted this to be a 10 digit number so we can differentiate between V1 and V2 reference numbers on FAA. I gave it a go and had to start from 1111111111.

File: src/Shared/Recruit.Vacancies.Client/Infrastructure/SequenceStore/MongoSequenceStore.cs [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+        public MongoSequenceStore(IOptions<MongoDbConnectionDetails> details)
+            : base(Database, Collection, details)
+        {
+        }
+
+        public async Task<long> GenerateAsync()
+        {
+            var collection = GetCollection<Sequence>();
+            var filter = Builders<Sequence>.Filter.Eq(x => x.Name, VacancyNumberSequenceName);
```
Comment (sachpatel) [inline src/Shared/Recruit.Vacancies.Client/Infrastructure/SequenceStore/MongoSequenceStore.cs]: Could the Id just be the name? Similar to what has been done in some of the queryviews such as for Dashboard and EditVacancyInfo

**Comment by leewadhams** [comment]
Have updated confluence to mirror change to sequence naming

**Pull request merged**